src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs

```diff
@@ -0,0 +1,1336 @@
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x3").WithLocation(19, 9),
+                // (20,9): error CS0165: Use of unassigned local variable 'y3'
+                //         y3.ToString(); // 6
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y3").WithArguments("y3").WithLocation(20, 9),
+                // (27,9): error CS0165: Use of unassigned local variable 'y4'
+                //         y4.ToString(); // 7
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y4").WithArguments("y4").WithLocation(27, 9));
+        }
+
+        [Fact]
+        public void UnconstrainedTypeParameter_PatternMatching()
+        {
+            var source =
+@"
+class C
+{
+    static void F1<T>(object o, T tin)
+    {
+        if (o is T t1)
+        {
+            t1.ToString();
+        }
+        else
+        {
+            t1 = default; // 1
+        }
+
+        t1.ToString(); // 2
+
+        if (!(o is T t2))
+        {
+            t2 = tin;
+        }
+        else
+        {
+            t2.ToString();
+        }
+        t2.ToString(); // 3
+
+        if (!(o is T t3)) return;
+        t3.ToString();
+    }
+}
+";
+
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (12,18): warning CS8652: A default expression introduces a null value when 'T' is a non-nullable reference type.
+                //             t1 = default; // 1
+                Diagnostic(ErrorCode.WRN_DefaultExpressionMayIntroduceNullT, "default").WithArguments("T").WithLocation(12, 18),
+                // (15,9): warning CS8602: Possible dereference of a null reference.
+                //         t1.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t1").WithLocation(15, 9),
+                // (25,9): warning CS8602: Possible dereference of a null reference.
+                //         t2.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t2").WithLocation(25, 9)
+                );
+        }
+
+        [WorkItem(32503, "https://github.com/dotnet/roslyn/issues/32503")]
+        [Fact]
+        public void PatternDeclarationBreaksNullableAnalysis()
+        {
+            var source = @"
+#nullable enable
+class A { }
+class B : A
+{
+    A M()
+    {
+        var s = new A();
+        if (s is B b) {}
+        return s; 
+    }
+} 
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void InferenceWithITuplePattern()
+        {
+            var source = @"
+#nullable enable
+class A { }
+class B : A
+{
+    A M()
+    {
+        var s = new A();
+        if (s is B b) {}
+        return s; 
+    }
+} 
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void RecursivePatternNullInferenceWithDowncast_01()
```

Comment:
Please expand the coverage of recursive patterns. I don't see any tests involving learning from non-null patterns, oblivious inputs being switched on, etc.